### ISSUE TITLE
Dialyzer seems unhappy with certain parameter type usage

### ISSUE THREAD

**Issue by Qqwy** [Qqwy/elixir-type_check#95]
@Qqwy I found a case, where the bug (c.f. #85 / #93 ) seems to persist... 

```elixir
### this module works fine, because of `any` in spec: `@spec! from_db(any) :: t()`
defmodule Types.Ad do
  use TypeCheck
  defstruct id: nil, consumer_price: nil

  @type! t :: %Types.Ad{
           id: number(),
           consumer_price: number()
         }

  @spec! from_db(any) :: t()
  def from_db(row) do
    %Types.Ad{
      id: Map.get(row, "id")
    }
  end
end

### this fails, because of `map` in spec: `@spec! from_db(map) :: t()`
defmodule Types.Ad3 do
  use TypeCheck
  defstruct id: nil

  @type! t :: %Types.Ad3{
           id: number()
         }

  @spec! from_db(map) :: t()
  def from_db(row) do
    %Types.Ad3{}
  end
end
```

ERROR message:
```
lib/type_check/spec.ex:22:call_without_opaque
Function call without opaqueness type mismatch.

Call does not have expected term of type
  {{TypeCheck.Builtin.Map.t()
    | %{
        :__struct__ => atom(),
        :choices => [any()],
        :element_type => _,
        :element_types => [any()],
        :keypairs => [any()],
        :local => boolean(),
        :name => atom(),
        :range => map(),
        :type => _,
        :value => _
      }, atom(),
    %{
      :expected_length => non_neg_integer(),
      :expected_size => integer(),
      :index => integer(),
      :key => _,
      :keys => [any()],
      :problem => _,
      :problems => [any()]
    }, _}, nil | maybe_improper_list() | map()}
  | {TypeCheck.Builtin.Map.t()
     | %{
         :__struct__ => atom(),
         :choices => [any()],
         :element_type => _,
         :element_types => [any()],
         :keypairs => [any()],
         :local => boolean(),
         :name => atom(),
         :range => map(),
         :type => _,
         :value => _
       }, atom(),
     %{
       :expected_length => non_neg_integer(),
       :expected_size => integer(),
       :index => integer(),
       :key => _,
       :keys => [any()],
       :problem => _,
       :problems => [any()]
     }, _}
 (with opaque subterms) in the 1st position.

TypeCheck.TypeError.exception(
  {{%TypeCheck.Spec{
      :location => {<<_::464>>, 30},
      :name => :from_db,
      :param_types => [
        %TypeCheck.Builtin.Map{:key_type => map(), :value_type => map()},
        ...
      ],
      :return_type => %TypeCheck.Builtin.NamedType{
        :local => false,
        :name => <<_::88>>,
        :type => %TypeCheck.Builtin.FixedMap{:keypairs => [any(), ...]}
      }
    }, :param_error,
    %{
      :index => 0,
      :problem =>
        {%TypeCheck.Builtin.Map{
           :key_type => %TypeCheck.Builtin.Any{},
           :value_type => %TypeCheck.Builtin.Any{}
         }, :not_a_map, %{}, _}
    }, [any(), ...]}, [{:file, <<_::464>>} | {:line, 22}, ...]}
)

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

I guess there is still some fishiness going on. Hope this helps a bit to narrow down this issue.

_Originally posted by @mindreframer in https://github.com/Qqwy/elixir-type_check/issues/93#issuecomment-1021481481_

**Comment by mindreframer** [comment]
@Qqwy Thanks! I've made a small repo to reproduce it here: 

- https://github.com/mindreframer/typecheck-bug-reproduction/runs/4950631067?check_suite_focus=true

**Comment by Qqwy** [comment]
I have done more tests to find out what happens here.

I have tried annotating the `TypeCheck.TypeError` and `TypeCheck.Spec` modules with `@dialyzer :no_warn`, but the error will still show up.

This makes me believe that it is part of the code that is generated by the macro. However, Dialyzer's output seems to indicate that it is a problem in how `TypeCheck.TypeError.exception` is called.

_What_ the exact problem is with this call I still have not figured out by the way.

**Help fixing this issue is greatly appreciated!**

**Comment by Qqwy** [comment]
Ah... I managed to fix it.

Turns out that Dialyzer was inferring the type of `TypeCheck.TypeError.exception` (even though it is a callback of a behaviour so I had expected it to use the type from `Exception.exception` or leave it alone alltogether) based on the type of `TypeCheck.TypeError.DefaultFormatter.format` which is called internally (But Dialyzer was not complaining about _that_ call but rather about the call to `exception`...).

And _this_ type signature is where the problem lies: `TypeCheck.Spec.t()` was not included in the possibilities for `TypeCheck.TypeError.Formatter.problem_tuple()` yet.
